server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java

```diff
@@ -211,6 +212,31 @@ public static <E> List<E> appendToCopy(Collection<E> collection, E element) {
         return Collections.unmodifiableList(Arrays.asList(array));
     }
 
+    /**
+     * Creates a copy of the given set and adds extra element.
+     *
+     * @param set     set to copy
+     * @param element element to add
+     */
+    public static <E> Set<E> addToCopy(Set<E> set, E element) {
+        var copy = new HashSet<E>(set);
+        copy.add(element);
+        return Collections.unmodifiableSet(copy);
+    }
+
+    /**
+     * Creates a copy of the given set and adds extra elements.
+     *
+     * @param set      set to copy
+     * @param elements elements to add
+     */
+    @SuppressWarnings("unchecked")
+    public static <E> Set<E> addToCopy(Set<E> set, E... elements) {
+        var copy = new HashSet<>(set);
+        copy.addAll(Arrays.asList(elements));
```

Comment:
Could the `copy` be sized to fit both the set and array?

Reply:
I updated the pr using the approach suggested by @arteam (contacting to streams). That should handle sizing